src/kvp.rs

```diff
@@ -0,0 +1,684 @@
+    /// * `ctx` - The current tracing context, used to access the span's metadata and status.
+    fn on_close(&self, id: Id, ctx: TracingContext<S>) {
+        if let Some(span) = ctx.span(&id) {
+            let span_status = span
+                .extensions()
+                .get::<SpanStatus>()
+                .copied()
+                .unwrap_or(SpanStatus::Success);
+
+            let end_time = SystemTime::now();
+
+            let span_context = span.metadata();
+            let span_id = Uuid::new_v4();
+
+            if let Some(start_instant) = span.extensions().get::<MyInstant>() {
+                let elapsed = start_instant.elapsed();
+
+                let start_time =
+                    end_time.checked_sub(elapsed).unwrap_or(UNIX_EPOCH);
+
+                let start_time_dt = DateTime::<Utc>::from(
+                    UNIX_EPOCH
+                        + start_time
+                            .duration_since(UNIX_EPOCH)
+                            .unwrap_or_default(),
+                )
+                .format("%Y-%m-%dT%H:%M:%S%.3fZ");
+
+                let end_time_dt = DateTime::<Utc>::from(
+                    UNIX_EPOCH
+                        + end_time
+                            .duration_since(UNIX_EPOCH)
+                            .unwrap_or_default(),
+                )
+                .format("%Y-%m-%dT%H:%M:%S%.3fZ");
+
+                let event_value = format!(
+                    "Start: {} | End: {} | Status: {}",
+                    start_time_dt, end_time_dt, span_status
+                );
+
+                self.handle_kvp_operation(
+                    span_status.level(),
+                    span_context.name(),
+                    &span_id.to_string(),
+                    &event_value,
+                );
+            }
+        }
+    }
+}
+
+/// Generates a unique event key by combining the event level, name, and span ID.
+///
+/// # Arguments
+/// * `event_level` - The logging level (e.g., "INFO", "DEBUG").
+/// * `event_name` - The name of the event.
+/// * `span_id` - A unique identifier for the span.
+fn generate_event_key(
+    event_level: &str,
+    event_name: &str,
+    span_id: &str,
+) -> String {
+    format!(
+        "{}|{}|{}|{}",
+        EVENT_PREFIX, event_level, event_name, span_id
+    )
+}
+
+/// Encodes a key-value pair (KVP) into one or more byte slices. If the value
+/// exceeds the allowed size, it is split into multiple slices for encoding.
+/// This is used for logging events to a KVP file.
+///
+/// # Arguments
+/// * `key` - The key as a string slice.
+/// * `value` - The value associated with the key.
+fn encode_kvp_item(key: &str, value: &str) -> Vec<Vec<u8>> {
+    let key_bytes = key.as_bytes();
+    let value_bytes = value.as_bytes();
+
+    let key_len = key_bytes.len().min(HV_KVP_EXCHANGE_MAX_KEY_SIZE);
+    let mut key_buf = vec![0u8; HV_KVP_EXCHANGE_MAX_KEY_SIZE];
+    key_buf[..key_len].copy_from_slice(&key_bytes[..key_len]);
```

Comment:
Thank you for pointing this out—great catch! I wasn’t fully considering the potential for a panic when indexing out of bounds; it’s definitely a subtle but important Rust gotcha. I'll make a note of this to avoid similar issues in the future!! 

Regarding the zero-padding, it is required due to how Hyper-V processes the telemetry. Hyper-V has a minimum length requirement for each entry, so when the entry length is below that threshold (which is often the case), we need to zero-fill to meet the requirement. I’ll make sure this is documented clearly though as part of the change!